### PR TITLE
feat: support kwok in dev-up

### PR DESCRIPTION
**Pull request by limhawjia**
This PR supports using kwokctl for setting up the dev environment using the dev-up script.

### CODE REVIEW

**Comment by limhawjia** [comment]
To use kwokctl, set the environment variable `CLUSTER_PROVIDER=kwok` when running `hack/dev-up.sh`.

**Comment by limhawjia** [comment]
Requires kwok v0.2.0

**Comment by limhawjia** [comment]
Additional changes:
* Fix regression where cluster-join timeout is always 0
* Cluster kubeconfigs created by the script are now written to individual separate files

File: docs/quickstart.md [hunk 1]
```diff
@@ -66,10 +62,12 @@ $ make manager
 $ ./output/manager --create-crds-for-ftcs \
     --klog-logtostderr=false \
     --klog-log-file "./kubeadmiral.log" \
-    --kubeconfig "$HOME/.kube/kubeadmiral/kubeconfig.yaml" \
+    --kubeconfig "$HOME/.kube/kubeadmiral/kubeadmiral-host.yaml" \
```
Comment (SOF3) [inline docs/quickstart.md:65]: why do we have to repeat the word `kubeadmiral` in the path here?

Comment (limhawjia) [reply docs/quickstart.md:65]: the generated kubeconfig files use the cluster names for convenience

Comment (gary-lgy) [reply docs/quickstart.md:65]: The names should include the cluster names, but is there a need to repeat `kubeadmiral`?

File: hack/dev-up.sh [hunk 2]
```diff
@@ -20,29 +20,44 @@ set -o pipefail
...
 HOST_CLUSTER_NAME=${HOST_CLUSTER_NAME:-"kubeadmiral-host"}
 MEMBER_CLUSTER_NAME=${MEMBER_CLUSTER_NAME:-"kubeadmiral-member"}
 MANIFEST_DIR=${MANIFEST_DIR:-"${REPO_ROOT}/config/crds"}
 CONFIG_DIR=${CONFIG_DIR:-"${REPO_ROOT}/config/sample/host"}
 NUM_MEMBER_CLUSTERS=${NUM_MEMBER_CLUSTERS:-"3"}
-
-mkdir -p "$(dirname "${KUBECONFIG_PATH}")"
-
-kind delete clusters "${HOST_CLUSTER_NAME}" $(seq 1 "${NUM_MEMBER_CLUSTERS}" | xargs -I{} echo "${MEMBER_CLUSTER_NAME}-{}")
+export CLUSTER_PROVIDER=${CLUSTER_PROVIDER:-"kind"}
```
Comment (gary-lgy) [inline hack/dev-up.sh:29]: Why is `CLUSTER_PROVIDER` exported while the rest are not?

Comment (gary-lgy) [reply hack/dev-up.sh:29]: If the rest are not required to be exported, can we mark them `local` for more clarity?

File: hack/util.sh [hunk 3]
```diff
@@ -23,6 +23,37 @@ function util::create_host_cluster() {
...
+  kubectl --kubeconfig="${KUBECONFIG_PATH}" --context="${HOST_CONTEXT}" annotate clusterrolebinding -A --all $NOFED
+  kubectl --kubeconfig="${KUBECONFIG_PATH}" --context="${HOST_CONTEXT}" annotate svc -A --all $NOFED
+}
+
+function util::create_host_kind_cluster() {
+  local HOST_CLUSTER_NAME=${1}
+  local KUBECONFIG_PATH=${2}
 
   local KIND_CONFIG
   KIND_CONFIG=$(mktemp)
```
Comment (gary-lgy) [inline hack/util.sh:59]: nit: creating a temp file may not be necessary here. 

Maybe one of the following might work?

```bash
KIND_CONFIG=(cat <<EOF
...
EOF)

kind create cluster --config <(echo "$KIND_CONFIG") ......
```

or

```bash
kind create cluster --config /dev/stdin ...... << EOF
...
EOF 
```

File: hack/util.sh [hunk 4]
```diff
@@ -44,50 +75,132 @@ nodes:
...
   rm "${KIND_CONFIG}"
 }
 
+function util::create_host_kwok_cluster() {
+  local HOST_CLUSTER_NAME=${1}
+  local KUBECONFIG_PATH=${2}
+
+  local KWOK_CONFIG
+  KWOK_CONFIG=$(mktemp)
+  APISERVER_PORT=$(shuf -i 30000-40000 -n 1)
```
Comment (gary-lgy) [inline hack/util.sh:88]: Why did we not choose to use `$RANDOM`?

File: hack/util.sh [hunk 5]
```diff
@@ -44,50 +75,132 @@ nodes:
...
-
-  local FCLUSTER FCLUSTER_SECRET
-  FCLUSTER=$(mktemp)
-  FCLUSTER_SECRET=$(mktemp)
-
-  cat <<EOF > "${FCLUSTER}"
+  local HOST_CONTEXT MEMBER_API_ENDPOINT MEMBER_CA MEMBER_CERT MEMBER_KEY
+  if [[ $CLUSTER_PROVIDER == "kind" ]]; then
+    HOST_CONTEXT="kind-${HOST_CLUSTER_NAME}"
+    MEMBER_API_ENDPOINT=$(kind get kubeconfig --name="${MEMBER_CLUSTER_NAME}" | grep 'server:' | awk '{ print $2 }')
```
Comment (gary-lgy) [inline hack/util.sh:176]: Why don't we get these from the kubeconfig directly?

File: docs/quickstart.md (hunk 1 above)
Comment (limhawjia) [reply docs/quickstart.md:65]: are you suggesting naming the clusters without `kubeadmiral` in them? or not organizing the kubeconfigs in a `kubeadmiral` directory?

File: cmd/controller-manager/app/util.go [hunk 6]
```diff
@@ -145,6 +145,7 @@ func createControllerContext(opts *options.Options) (*controllercontext.Context,
 func getComponentConfig(opts *options.Options) (*controllercontext.ComponentConfig, error) {
 	componentConfig := &controllercontext.ComponentConfig{
 		FederatedTypeConfigCreateCRDsForFTCs: opts.CreateCRDsForFTCs,
+		ClusterJoinTimeout:                   opts.ClusterJoinTimeout,
```
Comment (gary-lgy) [inline cmd/controller-manager/app/util.go:148]: This should already be in main. Rmb to rebase.

File: hack/dev-up.sh (hunk 2 above)
Comment (limhawjia) [reply hack/dev-up.sh:29]: oops, this does not need to be exported

File: hack/util.sh (hunk 3 above)
Comment (limhawjia) [reply hack/util.sh:59]: just tested and the first approach seems to work, will update the script

File: hack/util.sh (hunk 4 above)
Comment (limhawjia) [reply hack/util.sh]: good question 🙂, i shall switch to using `$RANDOM` for portability

File: docs/quickstart.md (hunk 1 above)
Comment (gary-lgy) [reply docs/quickstart.md:65]: The former. But given that the file name follows directly from the kind cluster name, let's leave it be for simplicity.

File: hack/util.sh (hunk 5 above)
Comment (gary-lgy) [reply hack/util.sh:169]: We wouldn't need to maintain 2 sets of logic if getting from kubeconfig directly. But not a big issue.

**Review by gary-lgy — APPROVED** [review]

**Pull request merged**